File: exercism/python/triangle/triangle.py
"""Triangle classification helpers for the Exercism triangle exercise."""

ISOSCELES_MAX_UNIQUE_SIDES = 2
SCALENE_UNIQUE_SIDES = 3


def is_triangle(sides) -> bool:
    """Return True when provided sides can form a non-degenerate triangle."""
    if not all_sides_exists(sides):
        return False

    first_side, second_side, third_side = sides
    return (
        first_side + second_side > third_side
        and second_side + third_side > first_side
        and first_side + third_side > second_side
    )


def all_sides_exists(sides) -> bool:
    """Return True when all sides have a positive, non-zero length."""
    return all(side_length > 0 for side_length in sides)


def equilateral(sides) -> bool:
    """Return True when all sides are equal and the sides form a triangle."""
    return is_triangle(sides) and len(set(sides)) == 1


def isosceles(sides) -> bool:
    """Return True when at least two sides are equal and valid triangle rules pass."""
    return is_triangle(sides) and len(set(sides)) <= ISOSCELES_MAX_UNIQUE_SIDES


def scalene(sides) -> bool:
    """Return True when all sides are different and the sides form a triangle."""
    return is_triangle(sides) and len(set(sides)) == SCALENE_UNIQUE_SIDES

File: exercism/typescript/triangle/triangle.ts
export class Triangle {
  a: number
  b: number
  c: number

  constructor(...sides: number[]) {
    this.a = sides[0];
    this.b = sides[1];
    this.c = sides[2];
  }

  private checkZeroLengthSides(): boolean {
    return this.a === 0 || this.b === 0 || this.c === 0;
  }

  private isTriangle(): boolean {
    return this.a + this.b > this.c &&
      this.a + this.c > this.b &&
      this.b + this.c > this.a;
  }

  get isEquilateral(): boolean {
    if (this.checkZeroLengthSides() || !this.isTriangle()) {
      return false
    }
    return this.a === this.b && this.b === this.c && this.c === this.a;
  }

  get isIsosceles(): boolean {
    if (this.checkZeroLengthSides() || !this.isTriangle()) {
      return false
    }
    return this.a === this.b || this.b === this.c || this.c === this.a;
  }

  get isScalene(): boolean {
    if (this.checkZeroLengthSides() || !this.isTriangle()) {
      return false
    }
    return (this.a !== this.b) && (this.b !== this.c) && (this.c !== this.a);
  }
}
